x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/action/profile/TransportProfileHasPrivilegesActionTests.java

```diff
@@ -180,10 +189,86 @@ public void testNoProfileSubjectsFound() throws Exception {
+        for (int i = 0; i < indicesPrivileges.length; i++) {
+            indicesPrivileges[i] = RoleDescriptor.IndicesPrivileges.builder()
+                .privileges(randomFrom("read", "write"))
+                .indices(randomAlphaOfLengthBetween(2, 8))
+                .query(new BytesArray(randomAlphaOfLength(5)))
+                .build();
+        }
+
+        final ProfileHasPrivilegesRequest request = new ProfileHasPrivilegesRequest(
+            randomList(1, 100, () -> randomAlphaOfLengthBetween(4, 10)),
+            new AuthorizationEngine.PrivilegesToCheck(
+                new String[0],
+                indicesPrivileges,
+                new RoleDescriptor.ApplicationResourcePrivileges[0],
+                randomBoolean()
+            )
+        );
+
+        final PlainActionFuture<ProfileHasPrivilegesResponse> listener = new PlainActionFuture<>();
+        transportProfileHasPrivilegesAction.execute(mock(CancellableTask.class), request, listener);
+
+        final IllegalArgumentException ile = expectThrows(IllegalArgumentException.class, () -> listener.actionGet());
+        assertThat(ile, notNullValue());
+        assertThat(ile.getMessage(), containsString("may only check index privileges without any DLS query"));
+    }
+
+    @SuppressWarnings("unchecked")
+    public void testCancellation() throws Exception {
+        final List<String> profileUids = new ArrayList<>(new HashSet<>(randomList(1, 5, () -> randomAlphaOfLengthBetween(5, 10))));
+        final ProfileHasPrivilegesRequest request = new ProfileHasPrivilegesRequest(profileUids, randomValidPrivilegesToCheckRequest());
+        doAnswer(invocation -> {
+            Collection<String> uidsArg = (Collection<String>) invocation.getArguments()[0];
+            Map<String, Subject> profileUidToSubject = new HashMap<>();
+            for (String uid : uidsArg) {
+                profileUidToSubject.put(uid, new Subject(new User("user_for_profile_" + uid), mock(Authentication.RealmRef.class)));
+            }
+            final ActionListener<MultiProfileSubjectResponse> listener = (ActionListener<MultiProfileSubjectResponse>) invocation
+                .getArguments()[1];
+            listener.onResponse(new MultiProfileSubjectResponse(profileUidToSubject, Set.of()));
+            return null;
+        }).when(profileService).getProfileSubjects(anyCollection(), anyActionListener());
+
+        doAnswer(invocation -> {
+            final ActionListener<Collection<ApplicationPrivilegeDescriptor>> listener = (ActionListener<
+                Collection<ApplicationPrivilegeDescriptor>>) invocation.getArguments()[2];
+            listener.onResponse(List.of());
+            return null;
+        }).when(nativePrivilegeStore).getPrivileges(anyCollection(), any(), anyActionListener());
+
+        final int partialResultSize = randomIntBetween(1, profileUids.size());
```

Comment:
You mean that the cancellation logic is not exercised in that case, right?

It's probably not ideal from a test design perspective, but it's also not completely irrelevant because, if the API supports partial results, there's still this case where the task is cancelled after the code last check the cancel status before returning a non-error response.
IMO it meets the bar for a randomization case.

PS: I initially wrote this test to mock the `CancellableTask#isCancelled`, but that is a final method, so I recurred to the currently proposed alternative (so this test makes it look like it is testing that every partial result is possible, but the goal is to cover that the task can be canceled at any point).